### ISSUE TITLE
thread safety of dgemv/m in openblas 0.3.3/2?

### ISSUE THREAD

**Issue by simonnwood** [xianyi/OpenBLAS#1761]
I maintain an R package (https://cran.r-project.org/web/packages/mgcv/) which has started producing test errors on a fedora machine which uses a single threaded OpenBLAS 0.3.2. My failing code is using openMP for parallelization and on digging I think that the issue is a thread safety problem with openBLAS 0.3.2 and 0.3.3. The problem does not occur in 0.2.20. 

In a nutshell I get substantially different answers when I use dgmev (of dgemm) in parallel rather than sequentially. It does not appear that any memory which dgemv ought to be modifying overlaps between threads.  

The code below is a small reproducible example. It (pointlessly) does a matrix multiplication by looping over columns and calling dgemv each time. The loop can optionally be parallelized using openmp in which case dgemv is called from parallel threads. The different threads modify non-overlapping parts of the matrix C. The multiplication is called once using parallel computation and once in single thread mode and the results compared. 

Compiling linking and running against the reference BLAS I get identical results when single or multi-threading (0 difference). The same is true for OpenBLAS 0.2.20. But for OpenBLAS  0.3.2 or 0.3.3 the results are substantially different between the two cases (difference around 1e8 for max abs truth of about 1e9) . 

```#include <stdlib.h>
#include <stdio.h>
#include <math.h>
#include <omp.h>
#include <cblas.h>

void mult(double *A,double *B,double *C,int m,int n,int nt) {
/* pointless C=A'B code using openmp. A and B are m by n. C is n by n. */
  int i,j;
#pragma omp parallel for private(i) num_threads(nt)
  for (i=0;i<n;i++) {
    cblas_dgemv(CblasColMajor, CblasTrans,m,n,1,A,m,B+i*m,1,0,C+i*n,1);
  }
}

void main() {
  double *A,*B,*C,*D,diff=0.0,x,max=0.0;
  int n = 200,m=300,i;
  A = (double *)calloc((size_t)n*m,sizeof(double));
  B = (double *)calloc((size_t)n*m,sizeof(double));
  C = (double *)calloc((size_t)n*n,sizeof(double));
  D = (double *)calloc((size_t)n*n,sizeof(double));
  for (i=0;i<n*m;i++) {A[i] = i*0.01;B[i]= n*m/(i*.04+.1);}
  printf("4 threads...\n");
  mult(A,B,C,m,n,4);        
  printf("done\n1 thread...\n");
  mult(A,B,D,m,n,1);
  printf("done\n");
  for (i=0;i<n*n;i++,D++,C++) { /* compare C & D which should be identical */
    x = fabs(*D-*C); if (x>diff) diff=x;
    if (*D>max) max = *D;
  }
  printf("Max |D| = %g\nMax abs diff = %g\n",max,diff);
}
```
My test set up is an Intel  i5-6300U CPU, Ubuntu 16.04,  gcc 5.4.0. OpenBLAS was built using 
```
make USE_THREAD=0 NO_LAPACK=1
```
Compilation and execution were 
```
gcc problem.c -I /opt/OpenBLAS/include/ -L /opt/OpenBLAS/lib/ -lopenblas -fopenmp
LD_LIBRARY_PATH=/opt/OpenBLAS/lib ./a.out
```
The original R package code called the Fortran BLAS routines directly, rather than using the cblas versions, and I got the same problem when I replaced dgemv calls with equivalent dgemm calls.

**Comment by martin-frbg** [comment]
Whatever it was must have happened between 0.3.0 and 0.3.1, I will try to find time for a proper `git bisect` later today.

**Comment by martin-frbg** [comment]
git bisect gives a399d00 as the first bad commit, which is #1625 (part of the recent rewrite of the thread memory initialization). You should not see this with 0.3.3 unless you build with the new option USE_TLS to choose this new and still unexpectedly fragile code - unfortunately it seems that my commit that introduced this option had USE_TLS=1 as the default in Makefile.rule (Quite contrary to
my intention to provide the proven if slower version as a safe default.)
Indeed commenting out that option makes your code run with "Max abs diff = 0" again. 
At least now we know that the new code may pose a problem for a non-threaded OpenBLAS as well @oon3m0oo

**Comment by brada4** [comment]
I dont think it is official R way to link some random cblas from outside. You can get Rblas exports from R_ext.h 
Especially rblas.so is already linked in process image providing same fortran double precision blas api symbols (dgem?_) that your explicitly linked in openblas is trying to show out. 
Could you try recordoing with 'perf' to ensure it was openblas being called and never ever Rblas.dll ?

**Comment by martin-frbg** [comment]
Note that the test case does not involve R at all, and the problem was reproducible and already tracked down to a single commit...

**Comment by simonnwood** [comment]
Thanks for looking at this so quickly. Just to confirm - commenting out USE_TLS=1 from Makefile.rule and re-building 0.3.3 works for me. The above test code runs correctly and so does the original code in my R package. 

btw: as I understand it 0.3.2 is the default BLAS on Fedora (which is why this showed up on one of the R test machines) - is it worth letting the Fedora people know?

**Comment by brada4** [comment]
Fedoras and epel testing is 0.3.3 and if that is broken too your feedback is valuable over there....

**Comment by brada4** [comment]
PS luckily epel is at 0.3.0 ....

**Comment by martin-frbg** [comment]
Fedora maintainer is taking part in discussion of 1735 already. 
I spent some time last night trying to debug this, but did not get anywhere. Probably the quickest (kludgy interim) solution would be simply to use the "old" version of memory.c when USE_THREAD=0, even if USE_TLS is set. (As @oon3m0oo 's work is intended to speed up the multithreaded cases, I think no performance should be lost if non-multithreaded code just uses the old and proven paths. Having both new and old code with a big `#ifdef' around them in memory.c is ugly of course)

**Comment by martin-frbg** [comment]
Closing as 0.3.4 is out with the fixes.
